crates/typst/src/export/render.rs

```diff
@@ -23,18 +24,24 @@ use crate::image::{Image, ImageKind, RasterFormat};
 ///
 /// This renders the frame at the given number of pixels per point and returns
 /// the resulting `tiny-skia` pixel buffer.
-pub fn render(frame: &Frame, pixel_per_pt: f32, fill: Color) -> sk::Pixmap {
-    let size = frame.size();
-    let pxw = (pixel_per_pt * size.x.to_f32()).round().max(1.0) as u32;
-    let pxh = (pixel_per_pt * size.y.to_f32()).round().max(1.0) as u32;
+pub fn render(frame: &Frame, pixel_per_pt: f32, fill: Color) -> Arc<sk::Pixmap> {
+    #[comemo::memoize]
+    fn cached(frame: &Frame, pixel_per_pt: u32, fill: Color) -> Arc<sk::Pixmap> {
```

Comment:
Here, I am very concered about memory usage. This will keep the pixmaps for all pages for the last 10-30 compilations. Assume a thirty page document with a lot of changes:

- 4 bytes per pixel
- 1200 * 900 page size (could be much larger when zoomed in)
- 30 pages as an example

=> 4 * 1200 * 900 * 30 bytes = 123 MB for all pages

For a big document and large changes, just the pages might eat up a GB or more. Especially in the web app that has only 2 GB available, that's a huge amount.

I think it should be up the caller of this function to retain pixel buffers they deem reusable. They can compare the frames themselves. `typst watch` could do this and the web app can decide for itself (it works differently anyway because it only renders visible pages).

Reply:
That's very true, another, perhaps cleaner idea I had was to cache manually at the I/O level (so caching around the file write) by taking the hashes of the pages manually and seeing if they change. That means we don't get caching for pages that change number (pagebreaks), but at least when editing one page, it does not invalidate all existing ones.

Reply:
That seems like a good solution.